### PR TITLE
JIRA# ROS-150: Implement C++ union semantics for Location

### PR DESCRIPTION
**Pull request by ghost**
There are some changes that involve indenting large blocks, so I recommend adding ?w=1 to the comparison page to get a nicer diff.

### CODE REVIEW

File: rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala
Comment (tymm) [inline rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Underscores in names are discouraged in Scala. See:  https://docs.scala-lang.org/style/naming-conventions.html

Comment (tymm) [reply rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Maybe we should bundle these vals into `object CtxtRegName`?

File: rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala
Comment (tymm) [inline rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Should be `AddrLevelSize` and `AddrOffsetSize`.

File: rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala
Comment (tymm) [inline rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: `sign` and `span` do not just get passed on.

**Review by tymm — CHANGES_REQUESTED** [review]
Four minor issues. See comments.

File: rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala
Comment (ghost) [reply rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Done

File: rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala
Comment (ghost) [reply rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Fixed

File: rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala
Comment (ghost) [reply rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Fixed

**Comment by ghost** [comment]
Thanks for the careful review.  I think I've now addressed all your points.

File: rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala
Comment (tymm) [inline rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Should be `BitField00SpanSize`.

Comment (tymm) [reply rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Without really thinking it through: Seems wrong since in the C++ implementation it is either `0` or `1`.

**Review by tymm — CHANGES_REQUESTED** [review]
Please do not merge in between reviews. Makes it much harder to review.

File: rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala
Comment (ghost) [reply rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Sorry about the merge; since I branched this one from dev-stay-location2, which was merged into dev, but wanted it to eventually get into dev itself, I wasn't sure how to go about that.  I switched the source of the comparison to dev, but also made a mistake with git remote that I had to fix.

Regarding line 389, the C++ says
```SET_BITFIELD00_SPAN(loc, ( span == (1 << BitField00SpanSize) ? 0 : span));```
which is the same as `span % (1 << BitField00SpanSize)` for the range 1 to (1 << BitField00SpanSize).

Comment (tymm) [reply rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala]: Ah sorry, I see: we are checking for that range above.
I'll approve once you change line `389` to `BitField00SpanSize`.

**Review by tymm — APPROVED** [review]

**Pull request merged**
